### PR TITLE
feat: get-products

### PR DESCRIPTION
**Pull request by mufidat3250**
## What does the PR do?

 Developed an API endpoint to list products with pagination support.

- [x] get all product.
- [x] products pagination

How should this be manually tested?

1. Make a Get request to the following endpoint 
   1. `\api\v1\products` 
    2. `\api\v1\products\?limit=10&page=1` 
2. The following payload must be put in place:

```json
Authorization: Bearer <access_token>
```
Successful Response

On a successful product list, the API should return a 200 get status code.

```json
{
  "success": true,
  "message": "Product retrieved successfully",
  "products": [
    {
      "name": "Product 1",
      "price": 29.99,
      "category": "Electronics"
    },
    {
      "name": "Product 2",
      "price": 19.99,
      "category": "Books"
    },
    ...
  ],
  "pagination": {
    "totalItems": 100,
    "totalPages": 10,
    "currentPage": 1
  },
"status_code": 200
}
```
success

Authentication Error:

When the current user is not properly authenticated to change the organization detail:

```json
{
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}
```
error

Task

- [x] Make a get request for product.
- [x] Make a pagination query
Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/97)

### CODE REVIEW

File: src/modules/product-category/product-category.controller.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { ProductCategoryService } from './product-category.service';
+import { CreateProductCategoryDto } from './dto/create-product-category.dto';
+import { UpdateProductCategoryDto } from './dto/update-product-category.dto';
+import { ApiTags, ApiBody, ApiParam } from '@nestjs/swagger';
+import { AuthGuard } from 'src/guards/auth.guard';
+import { skipAuth } from 'src/helpers/skipAuth';
+@ApiTags('products/categories')
+@UseGuards(AuthGuard)
+@skipAuth()
+@Controller('products/categories')
```
Comment (davidolanrewaju) [inline src/modules/product-category/product-category.controller.ts:12]: Remove the @skipAuth()  and @UseGuards() decorators

Comment (mufidat3250) [reply src/modules/product-category/product-category.controller.ts:7]: Done

**Review by davidolanrewaju — APPROVED** [review]

File: src/modules/product-category/product-category.controller.ts (hunk 1 above)
Comment (mufidat3250) [reply src/modules/product-category/product-category.controller.ts:7]: I have effected the changes

**Comment by CFCIfe** [comment]
| feat/get-products

to

| feat: get-products

... in your PR title.

File: src/modules/invite/tests/invite.service.spec.ts [hunk 2]
```diff
@@ -45,7 +45,7 @@ describe('InviteService', () => {
       time_left: 5,
       secret: 'secret',
       is_2fa_enabled: true,
-      testimonials: [],
+      products: []
```
Comment (CFCIfe) [inline src/modules/invite/tests/invite.service.spec.ts:48]: This is why your tests are failing.

Add the `testimonials: [],`

File: src/main.ts [hunk 3]
```diff
@@ -29,6 +30,7 @@ async function bootstrap() {
 
   app.enable('trust proxy');
   app.useLogger(logger);
+  app.useGlobalPipes(new ValidationPipe());
```
Comment (CFCIfe) [inline src/main.ts:33]: This can be taken off. We have this handled in the `app.module.ts` files

```
useFactory: () =>
        new ValidationPipe({
          whitelist: true,
          forbidNonWhitelisted: true,
        }),
```

File: src/main.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ import { DataSource } from 'typeorm';
 import { AppModule } from './app.module';
 import { initializeDataSource } from './database/data-source';
 import { SeedingService } from './database/seeding/seeding.service';
+import { ValidationPipe } from '@nestjs/common';
```
Comment (CFCIfe) [inline src/main.ts:10]: ...we don't need this here

File: src/modules/product-category/dto/create-product-category.dto.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  description?: string;
+
+  @ApiProperty({
+    description: 'Slug for the product category',
+    example: 'electronics',
+    required: false,
+  })
+  @IsNotEmpty()
+  @IsString()
+  slug?: string;
```
Comment (CFCIfe) [inline src/modules/product-category/dto/create-product-category.dto.ts:29]: If `description` is optional but, when provided, must be a non-empty string, won't using `@IsNotEmpty()` and `@IsString()` together not cause issues if description is not provided?

Please properly test all edge cases as well as for `slug` :)

File: src/modules/product-category/entities/product-category.entity.ts
Comment (CFCIfe) [inline src/modules/product-category/entities/product-category.entity.ts:1]: The dB diagram requires for slug to be stored.

image

File: src/modules/product-category/product-category.controller.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+
+import { Controller,Post, Body} from '@nestjs/common';
+import { ProductCategoryService } from './product-category.service';
+import { CreateProductCategoryDto } from './dto/create-product-category.dto';
+import { ApiTags, ApiBody} from '@nestjs/swagger';
+@ApiTags('products/categories')
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.controller.ts:6]:
```suggestion
@ApiTags('products-categories')
```

File: src/modules/product-category/product-category.controller.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+@Controller('products/categories')
+
+export class ProductCategoryController {
+  constructor(private readonly categoryService: ProductCategoryService) {}
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.controller.ts:10]:
```suggestion
constructor(private readonly productCategoryService: ProductCategoryService) {}
```

File: src/modules/product-category/product-category.controller.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  
+  @Post()
+  @ApiBody({ type: CreateProductCategoryDto })
+  async create(@Body() createCategoryDto: CreateProductCategoryDto) {
+    return {
+      status_code: 201,
+      category: await this.categoryService.create(createCategoryDto),
+    };
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.controller.ts:18]:
```suggestion
async createProductCategory(@Body() createProductCategoryDto: CreateProductCategoryDto) {
    return {
      status_code: 201,
      category: await this.categoryService.createProductCategory(createProductCategoryDto),
    };
```

File: src/modules/product-category/product-category.service.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+export class ProductCategoryService {
+  constructor(
+    @InjectRepository(ProductCategory)
+    private categoryRepository: Repository<ProductCategory>
+  ) {}
+
+  async create(createCategoryDto: CreateProductCategoryDto): Promise<ProductCategory> {
+    const category = this.categoryRepository.create(createCategoryDto);
+    return await this.categoryRepository.save(category);
+  }
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.service.ts:17]:
```suggestion
 async createProductCategory(createProductCategoryDto: CreateProductCategoryDto): Promise<ProductCategory> {
    const category = this.categoryRepository.create(createCategoryDto);
    return await this.categoryRepository.save(category);
  }
 ```

File: src/modules/product-category/product-category.service.ts
Comment (CFCIfe) [inline src/modules/product-category/product-category.service.ts:1]: Implement a `try...catch` block for dB calls

File: src/modules/products/dto/create-product.dto.ts
Comment (CFCIfe) [inline src/modules/products/dto/create-product.dto.ts:1]: Please work with this dB diagram. There are some columns not documented for.

image

File: src/modules/products/products.controller.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  @ApiResponse({ status: 401, description: 'Unauthorized. You must be authenticated to get list of products.' })
+  async findAllProducts(@Query('page') page: number, @Query('limit') limit: number) {
+    try {
+      return await this.productsService.findAllProducts(page, limit)
+    } catch (error) {
+      if(error instanceof NotFoundException){
+        throw new HttpException(error.getResponse(), error.getStatus());
+      }
+      throw error
+    }
```
Comment (CFCIfe) [inline src/modules/products/products.controller.ts:23]: Move the `try...catch` to your service file

File: src/modules/products/products.service.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+          status_code: 400,
+        });
+      }
+      if (page !== undefined && (isNaN(Number(page)) || Number(page) < 0)) {
+        throw new BadRequestException({
+          status: 'bad request',
+          message: 'Invalid query params passed',
+          status_code: 400,
+        });
+      }
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:33]: You should query the dB after the checks are done.

```suggestion
if (
  (limit !== undefined && (isNaN(limitNumber) || limitNumber < 0)) ||
  (page !== undefined && (isNaN(pageNumber) || pageNumber < 0))
) {
  throw new BadRequestException({
    status: 'bad request',
    message: 'Invalid query params passed',
    status_code: 400,
  });
}

const [products, totalCount] = await this.productRepository.findAndCount({
  take: limitNumber,
  skip: limitNumber * (pageNumber - 1),
});
```

File: src/modules/products/products.service.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      const totalPages = Math.ceil(totalCount / limit);
+      return {
+        success: true,
+        message: 'Product retrieved successfully',
+        products: products,
+        pagination: {
+          totalItems: totalCount,
+          totalPages: totalPages,
+          currentPage: page,
+        },
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:44]:
```suggestion
products,
        pagination: {
          totalItems: totalCount,
          totalPages,
          currentPage: page,
        },
```

File: src/modules/products/products.service.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        status_code: 200,
+      };
+    } catch (error) {
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:47]:
```suggestion
products,
        pagination: {
          totalItems: totalCount,
          totalPages,
          currentPage: page,
        },
        status_code: HttpStatus.OK,
        };
```

File: src/modules/products/products.service.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        success: true,
+        message: 'Product retrieved successfully',
+        products: products,
+        pagination: {
+          totalItems: totalCount,
+          totalPages: totalPages,
+          currentPage: page,
+        },
+        status_code: 200,
+      };
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:46]:
```suggestion
products,
        pagination: {
          totalItems: totalCount,
          totalPages,
          currentPage: page,
        },
```

**Review by CFCIfe — COMMENTED** [review]
Review done

File: src/modules/invite/tests/invite.service.spec.ts (hunk 2 above)
Comment (mufidat3250) [reply src/modules/invite/tests/invite.service.spec.ts:48]: Thanks for taking time out to review, I dont really understand this part of the review.

File: src/main.ts (hunk 3 above)
Comment (mufidat3250) [reply src/main.ts:33]: Done

File: src/main.ts (hunk 4 above)
Comment (mufidat3250) [reply src/main.ts:10]: Done

File: src/modules/product-category/dto/create-product-category.dto.ts (hunk 5 above)
Comment (mufidat3250) [reply src/modules/product-category/dto/create-product-category.dto.ts:29]: I get your point on this.
Removed @IsNotEmpty for both description and slog.

File: src/modules/product-category/product-category.controller.ts (hunk 6 above)
Comment (mufidat3250) [reply src/modules/product-category/product-category.controller.ts:6]: Done

File: src/modules/product-category/product-category.controller.ts (hunk 8 above)
Comment (mufidat3250) [reply src/modules/product-category/product-category.controller.ts:18]: Done

File: src/modules/product-category/product-category.service.ts (hunk 9 above)
Comment (mufidat3250) [reply src/modules/product-category/product-category.service.ts:17]: Done

File: src/modules/products/products.controller.ts (hunk 10 above)
Comment (mufidat3250) [reply src/modules/products/products.controller.ts:23]: changes effected

File: src/modules/products/products.service.ts (hunk 11 above)
Comment (mufidat3250) [reply src/modules/products/products.service.ts:33]: Done

File: src/modules/products/products.service.ts (hunk 14 above)
Comment (mufidat3250) [reply src/modules/products/products.service.ts:46]: Done

File: src/modules/invite/tests/invite.service.spec.ts (hunk 2 above)
Comment (mufidat3250) [reply src/modules/invite/tests/invite.service.spec.ts:48]: DONE

**Comment by mufidat3250** [comment]
@CFCIfe Done fixing 
Kindly review.

File: src/modules/product-category/entities/product-category.entity.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+// @Entity()
+// export class ProductCategory extends AbstractBaseEntity {
+// @Column({type:'text', unique: true})
+// name:string
+// @Column({type:'text', nullable: true})
+// description: string
+// @OneToMany(()=> Product, products => products.category)
+// product:Product
+// }
```
Comment (CFCIfe) [inline src/modules/product-category/entities/product-category.entity.ts:13]: Please remove comments.

File: src/modules/product-category/entities/product-category.entity.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+@Entity('categories')
+export class ProductCategory extends AbstractBaseEntity{
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
```
Comment (CFCIfe) [inline src/modules/product-category/entities/product-category.entity.ts:18]: You already have this when you `extend` AbstractBaseEntity. It can be removed

File: src/modules/product-category/entities/product-category.entity.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  name: string;
+
+  @Column('text')
+  description: string;
+
+  @Column('text')
+  slug: string;
+
+  @Column({ type: 'uuid', nullable: true })
+  parent_id: string;
```
Comment (CFCIfe) [inline src/modules/product-category/entities/product-category.entity.ts:30]: Sorry, what does this column do? If its your primary key, then it should be removed. If its a foreign key, please define the relationship.

File: src/modules/product-category/product-category.controller.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  constructor(private readonly productCategoryService: ProductCategoryService) {}
+  
+  @Post()
+  @ApiBody({ type: CreateProductCategoryDto })
+  async createProductCategory(@Body() createProductCategoryDto: CreateProductCategoryDto) {
+    return {
+      status_code: 201,
+      category: await this.productCategoryService.createProductCategory(createProductCategoryDto),
+    };
+  }
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.controller.ts:18]:
```suggestion
 async createProductCategory(@Body() createProductCategoryDto: CreateProductCategoryDto) {
    return await this.productCategoryService.createProductCategory(createProductCategoryDto)
  }
```

File: src/modules/product-category/product-category.service.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+  constructor(
+    @InjectRepository(ProductCategory)
+    private categoryRepository: Repository<ProductCategory>
+  ) {}
+
+  async createProductCategory(CreateProductCategoryDto: CreateProductCategoryDto): Promise<ProductCategory> {
+   try {
+    const category = this.categoryRepository.create(CreateProductCategoryDto);
+    return await this.categoryRepository.save(category);
+   } catch (error) {
```
Comment (CFCIfe) [inline src/modules/product-category/product-category.service.ts:18]:
```suggestion
 const category = this.categoryRepository.create(CreateProductCategoryDto);
 await this.categoryRepository.save(category);
 
 return {
     // The product category
 }
```

**Review by CFCIfe — COMMENTED** [review]
Additional comments.

File: src/modules/product-category/entities/product-category.entity.ts (hunk 15 above)
Comment (mufidat3250) [reply src/modules/product-category/entities/product-category.entity.ts:13]: Done

File: src/modules/product-category/entities/product-category.entity.ts (hunk 16 above)
Comment (mufidat3250) [reply src/modules/product-category/entities/product-category.entity.ts:18]: Done

File: src/modules/product-category/entities/product-category.entity.ts (hunk 17 above)
Comment (mufidat3250) [reply src/modules/product-category/entities/product-category.entity.ts:30]: I need to confirm from my other team member since i did not work on the product category directly

File: src/modules/product-category/product-category.service.ts (hunk 19 above)
Comment (mufidat3250) [reply src/modules/product-category/product-category.service.ts:18]: do you mean that i should assign a variable to  await this.categoryRepository.save(category); 
and then return it

File: src/modules/product-category/entities/product-category.entity.ts (hunk 17 above)
Comment (mufidat3250) [reply src/modules/product-category/entities/product-category.entity.ts:30]: just confirmed that it wont be needed for the stage of this application so i will be removing it.

**Comment by mufidat3250** [comment]
Done with the fixes

**Review by CFCIfe — COMMENTED** [review]
LGTM.

I'd need the approval to confirm those packages removed are not needed. I'd guess app startup will fail in CI/CD tho if they are.

I approve.

**Comment by mufidat3250** [comment]
I was told to always use npm that was why i removed the yarn lock 
i don't think that affect anything since they are both use to install packages
